js/runner.go

```diff
@@ -771,24 +776,35 @@ func (u *VU) runFn(
 		u.state.Tags.Set("iter", strconv.FormatInt(u.state.Iteration, 10))
 	}
 
-	defer func() {
-		if r := recover(); r != nil {
-			gojaStack := u.Runtime.CaptureCallStack(20, nil)
-			err = fmt.Errorf("a panic occurred in VU code but was caught: %s", r)
-			// TODO figure out how to use PanicLevel without panicing .. this might require changing
-			// the logger we use see
-			// https://github.com/sirupsen/logrus/issues/1028
-			// https://github.com/sirupsen/logrus/issues/993
-			b := new(bytes.Buffer)
-			for _, s := range gojaStack {
-				s.Write(b)
-			}
-			u.state.Logger.Log(logrus.ErrorLevel, "panic: ", r, "\n", string(debug.Stack()), "\nGoja stack:\n", b.String())
-		}
-	}()
-
 	startTime := time.Now()
-	v, err = fn(goja.Undefined(), args...) // Actually run the JS script
+
+	if u.moduleVUImpl.eventLoop == nil {
+		u.moduleVUImpl.eventLoop = newEventLoop(u.moduleVUImpl)
+	}
+	err = u.moduleVUImpl.eventLoop.start(func() (err error) {
+		// here the returned value purposefully shadows the external one as they can be different
+		defer func() {
+			if r := recover(); r != nil {
+				gojaStack := u.Runtime.CaptureCallStack(20, nil)
+				err = fmt.Errorf("a panic occurred in VU code but was caught: %s", r)
+				// TODO figure out how to use PanicLevel without panicing .. this might require changing
+				// the logger we use see
+				// https://github.com/sirupsen/logrus/issues/1028
+				// https://github.com/sirupsen/logrus/issues/993
+				b := new(bytes.Buffer)
+				for _, s := range gojaStack {
+					s.Write(b)
+				}
+				u.state.Logger.Log(logrus.ErrorLevel, "panic: ", r, "\n", string(debug.Stack()), "\nGoja stack:\n", b.String())
+			}
+		}()
+		v, err = fn(goja.Undefined(), args...) // Actually run the JS script
+		return err
+	})
+	if err != nil && isDefault {
```

Comment:
Why does `isDefault` matter here?

Reply:
Only when `isDefault` is true is `cancel` not nil, but I guess I should probably just check that `cancel` isn't nil :) 

Also on related note looking at this now again, we only cancel the context on error ... which likely isn't so great. 

Shall I : 
1. do nothing now? it works for the cases we care about and this code likely will change eitehr way
2. chance to check `cancel!=nil` instead - practically 0% change in what will happen hopefully :crossed_fingers: 
3. Do actually fix this and make the whole check just check if there is `cancel`  - this makes it so we always will cancel the context. Arguably what the code should've done but it might be a bit too close to the release to change this :thinking: . But still not that *big* of a change IMO

Reply:
I went to try to do `3` and to see what problems exist and found:
1. actually in all cases there is a `cancel` that can be provided so we can truly wait on setup,teardown and handleSummary as well. Whether that is that much of an improvement to do is a different thing
2. (which I remember thinking about but apparently never checked ?!?) a side effect of canceling the context is that the code below that checks if this was a full iterations or not will always say that *it wasn't* a full iteration. Moving that code before here fixes that.  So with the current code if there is an exception we will mark an iteration as not full, which is a change as well.

Reply:
Final analysis (correcting some mistakes from above):
With the current code in all cases the context gets canceled when an iteartion end or `setup` gets executed. This happens as I've added it for iterations and it was always the case for the other code.
The current code *specifically* for iterations (of the "default" function) will *on* error:
1. cancel the context *early*
2. wait for the event loop to be empty
this so we don't just keep making more and more promises in the event loop and then an exception abort the iteration, but k6 doesn't actually wait for that concurrent work wind down. But instead just starts a new iteration doing the same.
This means that this also changed that an iteartion that had an exception is now marked as *not* full.

Possible changes to the code:
1. check `cancel != nil` instead of `isDefault`
2. Move the `isFullIteration` logic *above* this code so that `isFullIteration` is true even on an excepiton - mostly to be compatible with the old version.
3. Make this behaviour also true for not default function

1 and 2 are pretty small and safe changes IMO. 
3. is a bit more ... risky

Reply:
I'm probably not understanding all the subtleties at play here, but my hunch is 1./2.